Lib/typing.py

```diff
@@ -170,8 +170,11 @@ def _type_check(arg, msg, is_argument=True, module=None, *, is_class=False):
             invalid_generic_forms += (Final,)
 
     arg = _type_convert(arg, module=module)
-    if (isinstance(arg, _GenericAlias) and
-            arg.__origin__ in invalid_generic_forms):
+    if (
+        (isinstance(arg, _GenericAlias) and
+            arg.__origin__ in invalid_generic_forms) or
+        (is_argument and arg == Final)  # Raw `Final` is not `_GenericAlias`
+    ):
```

Comment:
Nit but I prefer to make more complex `if` statements trivial to read. Plus, operators should go on the left side in multi-line expressions per PEP 8.

```suggestion
    is_invalid_generic = (
        isinstance(arg, _GenericAlias)
        and arg.__origin__ in invalid_generic_forms
    )
    is_invalid_bare_final = is_argument and arg is Final
    if is_invalid_generic or is_invalid_bare_final:
```